simulator/infra/ports/t/unit_test.cpp

```diff
@@ -219,3 +220,154 @@ TEST_CASE("Module: exclusion")
+            to_D = make_write_port<int>( "B_to_D", PORT_BW);
+            from_C = make_read_port<int>( "C_to_B", PORT_LATENCY);
+            from_D = make_read_port<int>( "D_to_B", PORT_LATENCY);
+        }
+    };
+    struct C : public Module {
+        std::unique_ptr<WritePort<int>> to_A;
+        std::unique_ptr<WritePort<int>> to_B;
+        std::unique_ptr<ReadPort<int>> from_A;
+        std::unique_ptr<ReadPort<int>> from_B;
+        explicit C( Module* root) : Module( root, "C")
+        {
+            to_A = make_write_port<int>( "C_to_A", PORT_BW);
+            to_B = make_write_port<int>( "C_to_B", PORT_BW);
+            from_A = make_read_port<int>( "A_to_C", PORT_LATENCY);
+            from_B = make_read_port<int>( "B_to_C", PORT_LATENCY);
+        }
+    };
+    struct D : public Module {
+        std::unique_ptr<WritePort<int>> to_A;
+        std::unique_ptr<WritePort<int>> to_B;
+        std::unique_ptr<ReadPort<int>> from_A;
+        std::unique_ptr<ReadPort<int>> from_B;
+        explicit D( Module* root) : Module( root, "D")
+        {
+            to_A = make_write_port<int>( "D_to_A", PORT_BW);
+            to_B = make_write_port<int>( "D_to_B", PORT_BW);
+            from_A = make_read_port<int>( "A_to_D", PORT_LATENCY);
+            from_B = make_read_port<int>( "B_to_D", PORT_LATENCY);
+        }
+    };
+    std::unique_ptr<A> a;
+    std::unique_ptr<B> b;
+    std::unique_ptr<C> c;
+    std::unique_ptr<D> d;
+    void modulemap_load( pt::ptree& pt_modulemap) const { modulemap_dumping( pt_modulemap); }
+    void portmap_load( pt::ptree& pt_portmap) const { portmap_dumping( pt_portmap); }
+    void modules_load( pt::ptree& pt_modules) const { module_dumping( pt_modules); }
+    void topology_load( pt::ptree& pt_topology) const { topology_dumping_impl( pt_topology); }
+    void topology_save() { topology_dumping(true); }
+    void topology_save(const std::string& filename) { topology_dumping(true, filename); }
+    bool check_if_dumps() const { return sout.enabled(); }
+    explicit SomeTopology()
+    {
+        /* Root
+         * | \
+         * A  C
+         * |   \ 
+         * B    D  
+         */
+        a = std::make_unique<A>( this);
+        c = std::make_unique<C>( this);
+        b = std::make_unique<B>( a.get());
+        d = std::make_unique<D>( c.get());
+        init_portmap();
+    }
+};
+
+TEST_CASE( "Topology: modulemap")
+{
+    std::string file_path = __FILE__;
+    boost::algorithm::replace_last( file_path, "unit_test.cpp", "modulemap_test.json");
+    SomeTopology t;
+    pt::ptree pt_exp_modulemap;
+    pt::ptree pt_modulemap;
+    CHECK_NOTHROW( read_json( file_path, pt_exp_modulemap));
+    CHECK_NOTHROW( t.modulemap_load( pt_modulemap));
+    CHECK( pt_modulemap == pt_exp_modulemap);
+}
+
+TEST_CASE( "Topology: portmap")
+{
+    std::string file_path = __FILE__;
+    boost::algorithm::replace_last( file_path, "unit_test.cpp", "portmap_test.json");
```

Comment:
Slightly overengineered here. We usually use path relative to CMakeLists, let's stick to it:
https://github.com/MIPT-ILab/mipt-mips/blob/16c098bb3a1dfd8f107d66633787d4794aaf9f22/simulator/func_sim/t/unit_test.cpp#L33